ingest/bin/apply-geolocation-rules

```diff
@@ -0,0 +1,226 @@
+
+
+            geolocation_rules[raw[0]][raw[1]][raw[2]][raw[3]] = annot
+
+    return geolocation_rules
+
+
+def get_annotated_geolocation(geolocation_rules, raw_geolocation, rule_traversal = None):
+    """
+    Gets the annotated geolocation for the *raw_geolocation* in the provided
+    *geolocation_rules*.
+
+    Recursivley traverses the *geolocation_rules* until we get the annotated
+    geolocation, which must be a Tuple. Returns `None` if there are no
+    applicable rules for the provided *raw_geolocation*.
+
+    Rules are applied in the order of region, country, division, location.
+    First checks the provided raw values for geolocation fields, then if there
+    are not matches, tries to use general rules marked with '*'.
+    """
+    # Always instantiate the rule traversal as an empty list if not provided,
+    # e.g. the first call of this recursive function
+    if rule_traversal is None:
+        rule_traversal = []
+
+    current_rules = geolocation_rules
+    # Traverse the geolocation rules based using the rule_traversal values
+    for field_value in rule_traversal:
+        current_rules = current_rules.get(field_value)
+        # If we hit `None`, then we know there are no matching rules, so stop the rule traversal
+        if current_rules is None:
+            break
+
+    # We've found the tuple of the annotated geolocation
+    if isinstance(current_rules, tuple):
+        return current_rules
+
+    # We've reach the next level of geolocation rules,
+    # so try to traverse the rules with the next target in raw_geolocation
+    if isinstance(current_rules, dict):
+        next_traveral_target = raw_geolocation[len(rule_traversal)]
+        rule_traversal.append(next_traveral_target)
+        return get_annotated_geolocation(geolocation_rules, raw_geolocation, rule_traversal)
+
+    # We did not find any matching rule for the last traversal target
+    if current_rules is None:
+        # If we've used all general rules and we still haven't found a match,
+        # then there are no applicable rules for this geolocation
+        if all(value == '*' for value in rule_traversal):
+          return None
+
+        # Find the index of the first of the consecutive '*' from the
+        # end of the rule_traversal
+        # [A, *, B, *] => first_consecutive_general_rule_index = 3
+        # [A, *, *, *] => first_consecutive_general_rule_index = 1
+        first_consecutive_general_rule_index = 0
+        for index, field_value in reversed(list(enumerate(rule_traversal))):
+            if field_value == '*':
+                first_consecutive_general_rule_index = index
+            else:
+                break
+
+        # Edit the raw value of the field directly before the first consecutive '*'
+        # in the rule traversal in hopes that by moving to a general rule on
+        # a higher level, we can find a matching rule.
+        rule_traversal[first_consecutive_general_rule_index - 1] = '*'
```

Comment:
@joverlee521 I'm a bit confused here, could you provide an example of what this is trying to do?

Reply:
Yeah, this might be hard to explain. I'll walk through an example.

Example rules:

|rule # | raw | annotated|
|-------------|--------|---------|
| 1 | a/b/c/d | 1/2/3/4 |
| 2 | a/\*/\*/\* | 1/\*/\*/\* |

A example case: `( a, b, c, e )`

Here is how `rule_traversal` will be edited per functional call:
1. [a]
2. [a, b]
3. [a, b, c] <- Up to this point, we are matching rule # 1
4. [a, b, c, e] <- This doesn't match any available rules, so we try to backfill with `*`
5. [a, b, c, * ] <- Still no matching rule, continue `*` backfill
6. [a, b, *, *] <- Still no matching rule, continue `*` backfill
7. [a, *, *, * ] <- This now matches rule # 2!

And of course as I explain this, I'm realizing there are cases where this logic would not work...I'll create a PR to update this tomorrow

Reply:
See https://github.com/nextstrain/monkeypox/pull/55 for the updated logic 🙏